src/NServiceBus.RavenDB/Timeouts/TimeoutPersister.cs

```diff
@@ -55,62 +49,59 @@ public TimeoutPersister()
                 results = new List<Tuple<string, DateTime>>();
             }
 
-            RavenQueryStatistics statistics;
+            // default return value for when no results are found
+            nextTimeToRunQuery = DateTime.UtcNow.AddMinutes(10);
 
             using (var session = DocumentStore.OpenSession())
             {
-                int totalCount, skipCount = 0;
+                var query = GetChunkQuery(session)
+                    .Where(t => t.Time > startSlice)
+                    .Select(t => new
+                    {
+                        t.Id,
+                        t.Time
+                    });
 
-                do
+                QueryHeaderInformation qhi;
+                using (var enumerator = session.Advanced.Stream(query, out qhi))
                 {
-                    if (CancellationRequested())
+                    if (qhi.TotalResults < 1000)
                     {
-                        return Enumerable.Empty<Tuple<string, DateTime>>();
+                        logger.DebugFormat("Retrieving {0} timeouts.", qhi.TotalResults);
                     }
-
-                    var query = GetChunkQuery(session);
-
-                    var dueTimeouts =
-                        query.Statistics(out statistics)
-                            .Where(t => t.Time >= startSlice)
-                            .Where(t => t.Time < now)
-                            .Skip(skipCount)
-                            .Select(t => new
-                            {
-                                t.Id,
-                                t.Time
-                            })
-                            .Take(maximumPageSize);
-
-                    foreach (var dueTimeout in dueTimeouts)
+                    else if (qhi.TotalResults < 10000)
                     {
-                        results.Add(new Tuple<string, DateTime>(dueTimeout.Id, dueTimeout.Time));
+                        logger.InfoFormat("Retrieving {0} timeouts. This could take some time.", qhi.TotalResults);
                     }
-
-                    if (CancellationRequested())
+                    else
                     {
-                        return Enumerable.Empty<Tuple<string, DateTime>>();
+                        logger.WarnFormat("Retrieving {0} timeouts. This could take quite some time.", qhi.TotalResults);
                     }
 
-                    skipCount = results.Count + statistics.SkippedResults;
-                    totalCount = statistics.TotalResults;
-                }
-                while (results.Count < totalCount);
+                    while (enumerator.MoveNext())
```

Comment:
Just a call with @ramonsmits on this.

if we have a milion expired timeouts, as we said, we are going to return a LOT of data. One option could be to say: if there are more than X timeouts in the query we just set a hard stop at Y and suggest to core to get back as soon as possible (since we have more). 

Obviously this has an interesting side effect if those milion expired timeouts expires all *at the same time* the next start slice will return the same timeouts as the previous one. The solution is to `orderby Time AND CreationDate` and use the creation date of the last timeout we return as the next start slice.

@DavidBoike thoughts?